File: findAllMissingNumbers.java
// Time Complexity : O(n), where n is the number of elements in the array
// Space Complexity : O(1), no extra space used
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
/**
 * 1. Iterate through the array and mark the index of the number as negative.
 * 2. Iterate through the array again and add the index of the positive numbers to the result list and return it.
 * 3. Positive numbers in the array indicate that the index of the number is missing.
 */

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        int len = nums.length;
        ArrayList<Integer> list = new ArrayList<>();

        for(int i=0; i<len; i++){
            int currNum = Math.abs(nums[i]);
            if(nums[currNum-1] > 0){
               nums[currNum-1] *= -1; 
            }
        }

        for(int i=0; i<len; i++){
            if(nums[i] > 0){
                list.add(i+1);
            }
        }

        return list;
    }
}

File: FindAllMissingNumbers.py
#Time Complexity : O(n)
#Space Complexity : o(1) (Since output array is counted as auxilary space)

#Did this code successfully run on Leetcode : Yes
#Three line explanation of solution in plain english: Since all values in the problem are between 1 to N there can be no negative numbers in the array, plus the number has to be between 1 to N. So, iterating through the array  if we come across a number we just make that numbered index negative (if it's the first time we came across that number, since there can be repetitions). And then finally we go through the array and see if the number at that index is negative or not, if it is positive then that means the number is not present in the array and thus we append to result array.

#Your code here along with comments explaining your approach

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        #optimized
        missing = []

        #All the numbers present, make that numbered index negative
        for i in range(len(nums)):
            #if number is not present
            if nums[abs(nums[i])-1] > 0:
                nums[abs(nums[i])-1] *= -1

        #find all numbers which are not present in the array
        for i in range(1, len(nums)+1):
            if nums[i-1] > 0:
                missing.append(i)

        return missing

        #using extra space
#         n = len(nums)
#         allNumbers = set(nums)
#         missing = []

#         for i in range(1,len(nums)+1):
#             if i not in allNumbers:
#                 missing.append(i)

#         return missing
